src/udp/connection_cookie.rs

```diff
@@ -0,0 +1,214 @@
+
+        assert_ne!(
+            make_test_cookie(Some(&test_remote_address_1)),
+            make_test_cookie(Some(&test_remote_address_2))
+        );
+
+        assert_ne!(
+            make_test_cookie(Some(&test_remote_address_1)),
+            make_test_cookie(Some(&test_remote_address_3))
+        );
+
+        assert_ne!(
+            make_test_cookie(Some(&test_remote_address_2)),
+            make_test_cookie(Some(&test_remote_address_3))
+        )
+    }
+
+    #[test]
+    fn it_should_make_different_cookies_for_the_next_period() {
+        let cookie_now = make_test_cookie(None);
+
+        StoppedClock::local_add(&get_period_length()).unwrap();
+
+        let cookie_next = make_test_cookie(None);
+
+        assert_ne!(cookie_now, cookie_next)
+    }
+
+    #[test]
+    fn it_cookies_should_be_valid_for_this_and_next_periods() {
+        let cookie_now = make_test_cookie(None);
+
+        check_connection_cookie(&get_test_socket_addr(), &cookie_now).unwrap();
+
+        StoppedClock::local_add(&get_period_length()).unwrap();
+
+        check_connection_cookie(&get_test_socket_addr(), &cookie_now).unwrap();
+    }
+
+    #[test]
+    fn it_cookies_should_be_valid_for_the_last_period() {
+        let cookie_now = make_test_cookie(None);
+
+        let last_period = get_last_valid_period_now();
+
+        StoppedClock::local_set(&detail::get_time_at_period_start(&(last_period)));
+
+        check_connection_cookie(&get_test_socket_addr(), &cookie_now).unwrap();
+    }
+
+    #[test]
+    #[should_panic]
+    fn it_cookies_should_be_not_valid_after_their_periods_has_expired() {
+        let cookie_now = make_test_cookie(None);
+
+        let last_period = get_last_valid_period_now();
+
+        StoppedClock::local_set(&detail::get_time_at_period_end(&(last_period)));
+
+        check_connection_cookie(&get_test_socket_addr(), &cookie_now).unwrap();
+    }
+}
+
+mod detail {
+    use std::collections::hash_map::DefaultHasher;
+    use std::hash::{Hash, Hasher};
+    use std::time::Duration;
+
+    use super::*;
+    use crate::protocol::clock::{DefaultClock, Time};
+    use crate::protocol::crypto::keys::seeds::{DefaultSeed, Seed};
+
+    pub(crate) type Period = u128;
+    pub(crate) type Periods = u64;
+
+    pub(crate) type Cookie = u64;
+    pub(crate) type Seconds = u64;
+
+    pub(crate) const PERIODS_IN_LIFETIME: Periods = 60;
+
+    pub(crate) const PERIOD_LENGTH: Seconds = 2;
+
+    pub(crate) fn get_period_now() -> Period {
+        DefaultClock::now_periods(&Duration::from_secs(PERIOD_LENGTH)).unwrap()
+    }
+
+    pub(crate) fn get_time_at_period_start(period: &Period) -> SinceUnixEpoch {
+        let periods_sec = (*period).saturating_mul(PERIOD_LENGTH as u128);
+        SinceUnixEpoch::from_secs(u64::try_from(periods_sec).unwrap())
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn get_time_at_period_end(period: &Period) -> SinceUnixEpoch {
+        get_time_at_period_start(&(period + 1))
+    }
+
```

Comment:
@da2ce7 could we extract a `Period` struct?:

```rust
pub struct Period(u128);

impl Period {
    pub fn get_current_period() { ... }
    pub get_start_time(&self) { ... }
    pub get_end_time(&self) { ... }
}
```

Reply:
Done :) I have extracted the period concept into the clock.